consensus/src/lib.rs

```diff
@@ -1,27 +1,624 @@
+                if witness_merkle_root != witness_merkle {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            Err(_merkle_error) => {
+                // TODO: Should we return additional error information?
+                return Err(BlockError::Unknown);
+            }
+        }
+
+        match &block.get_prev_block_id() {
+            Some(block_id) => {
+                let previous_block = self
+                    .db_tx
+                    .get_block_index(&Id::<Block>::from(block_id))?
+                    .ok_or(BlockError::NotFound)?;
+                // Time
+                let block_time = block.get_block_time();
+                if previous_block.get_block_time() > block_time {
+                    return Err(BlockError::Unknown);
+                }
+                if i64::from(block_time) > time::get() {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            None => (),
+        }
+
+        self.check_transactions(block)?;
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_consensus(&self, block: &Block) -> Result<(), BlockError> {
+        let _consensus_data = block.get_consensus_data();
+        // TODO: PoW is not in master at the moment =(
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_transactions(&self, _block: &Block) -> Result<(), BlockError> {
+        //TODO: Must check for duplicate inputs (see CVE-2018-17144)
+        //TODO: Size limits
+        //TODO: Check signatures
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_block(&self, block: &Block) -> Result<(), BlockError> {
+        self.check_consensus(block)?;
+        self.check_block_detail(block)?;
+        Ok(())
+    }
+
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> bool {
+        block_index.next_block_id.is_some()
+            || match self.db_tx.get_best_block_id().ok().flatten() {
+                Some(ref block_id) => block_index.get_block_id() == block_id,
+                None => false,
+            }
+    }
+
+    /// Mark new block as an orphan
+    /// Ok(()) - Added
+    /// Err(BlockError) - StorageFailure  
+    fn new_orphan_block(&mut self, block: Block) -> Result<(), BlockError> {
+        // If we have not the previous block we have to move it to OrphanBlocksPool, except if it genesis block
+        match (
+            block.get_prev_block_id(),
+            block.get_id() != self.chain_config.genesis_block().get_id(),
+        ) {
+            (Some(_), _) => (), // Not a genesis block and have the ancestor
+            (None, true) => {
+                // Not a genesis and have no the ancestor
+                self.orphan_blocks.add_block(block).map_err(|err| match err {
+                    OrphanAddError::BlockAlreadyInOrphanList(_) => BlockError::Orphan,
+                })?;
+            }
+            (None, false) => (), // genesis
+        }
```

Comment:
Not sure how other people see it, but I would find it easier to understand if the genesis block condition was flipped, i.e. change `!=` to `==`, and flip `true` and `false` in the patterns. I find myself doing more mental negations this way.